Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -477,57 +436,54 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
             autoreleasepool {
                 // Walk the node and extract external link references.
-                var externalLinksCollector = ExternalReferenceWalker(bundle: bundle)
-                externalLinksCollector.visit(object.semantic)
+                var externalLinksCollector = ExternalReferenceWalker(localBundleID: localBundleID)
+                externalLinksCollector.visit(semantic)
+
+                // Avoid any synchronization overhead if there are no references to add.
+                guard !externalLinksCollector.collectedExternalReferences.isEmpty else { return }
                 
                 // Add the link pairs to `collectedExternalLinks`.
-                results.append(contentsOf:
-                    externalLinksCollector.collectedExternalReferences.map { unresolved -> ExternalLinkResult in
-                        return ExternalLinkResult(unresolved: unresolved, targetLanguage: object.reference.sourceLanguage)
+                collectedExternalLinks.sync {
+                    for (bundleID, collectedLinks) in externalLinksCollector.collectedExternalReferences {
+                        $0[bundleID, default: []].formUnion(collectedLinks)
                     }
-                )
+                }
             }
         }
         
-        try Set(collectedExternalLinks).compactMap { externalLink -> (url: ValidatedURL, resolved: TopicReferenceResolutionResult)? in
-            guard let referenceBundleIdentifier = externalLink.unresolved.topicURL.components.host else {
-                assertionFailure("Should not hit this code path, url is verified to have an external bundle id.")
-                return nil
+        for (bundleID, collectedLinks) in collectedExternalLinks.sync({ $0 }) {
+            guard let externalResolver = externalDocumentationSources[bundleID] else {
+                continue
             }
-            
-            if let externalResolver = externalReferenceResolvers[referenceBundleIdentifier] {
-                let reference = externalResolver.resolve(.unresolved(externalLink.unresolved), sourceLanguage: externalLink.targetLanguage)
-                if case .success(let resolvedReference) = reference {
+            for externalLink in collectedLinks {
+                let unresolvedURL = externalLink.topicURL
+                let result = externalResolver.resolve(.unresolved(externalLink))
+                externallyResolvedLinks[unresolvedURL] = result
+                
+                if case .success(let resolvedReference) = result {
                     // Add the resolved entity to the documentation cache.
-                    if let externallyResolvedNode = try externalEntity(with: resolvedReference) {
-                        documentationCache[resolvedReference] = externallyResolvedNode
+                    if let externallyResolvedNode = externalEntity(with: resolvedReference) {
+                        externalCache[resolvedReference] = externallyResolvedNode
```

Comment:
This is the code that populates the cache with the successfully resolved external links. 

The call to `externalEntity(with:)` asks the `ExternalDocumentationSource` (previously called `ExternalReferenceResolver`) to create an entity for the already resolved reference. That entity is then added to the cache (via a mutating subscript) but there's no reading from the cache here.

There's also no updates of the `symbolIndex` happening here. The way that `externalCache.storage`, `externalCache.symbolIndex`, and `externallyResolvedLinks` relate to each other can be quite subtle. I'm not quite sure where it would be a good place to write it down... maybe just in the documentation for `DocumentationContext/externalCache` or `DocumentationContext/externallyResolvedLinks`🤔 

The `externalCache` and `externallyResolvedLinks` are built up in a two different steps. First, while the symbols from the SGFs are being registered, any USR in a relationships or declaration tokens that can't be found locally is attempted to be found via the `GlobalExternalSymbolResolver`. Here the input is just the USR. If the global symbol resolver can find that symbol it caches both the *USR -> Reference* in `externalCache.symbolIndex` and the *Reference -> Entity* in `externalCache.storage`.

Later, when the symbols have all been registered, any documentation extension files have already been associated with the symbols, and the combined markup has been parsed, the context gathers all the authored documentation links with external bundle identifiers and asks the corresponding `ExternalDocumentationSource` (previously called `ExternalReferenceResolver`) to resolve those links. Regardless if the link resolved successfully or not, the result is cached in `externallyResolvedLinks` so that the context doesn't attempt to re-resolve the unsuccessful links during rendering. If the link successfully resolved, the context asks for the entity matching that reference and caches that entity in `externalCache.storage`. At this point there's no point in adding any more USRs to `externalCache.symbolIndex` because all the external symbol references have already been encountered.

Reply:
+1 to writing how `externalCache.storage`, `externalCache.symbolIndex` relate down!
It took me a little while to figure out how they're supposed to be used together, I think partially because their names don't make that clear.

Reply:
I added a shorter version of this information in https://github.com/apple/swift-docc/pull/806/commits/3616630d4a293e45e885b0fd03a88baf8dc9d94b